### ISSUE TITLE
misleading error on Regex matching with malformed UTF-8

### ISSUE THREAD

**Issue by kolia** [JuliaLang/julia#40845]
```julia
julia> m = match(r"\s*[co]", "\xf6r")
ERROR: ReadOnlyMemoryError()
```

```julia
julia> m = match(r"\s*[r]", "\xf6r")
ERROR: ReadOnlyMemoryError()
```

but for slightly different regexes there is no error, and it is hard to predict when/why

```julia
julia> nothing == match(r"\s*[c]", "\xf6r")
true
```

Probably has something to do with

```julia
julia> "\xf6"[1]
'\xf6': Malformed UTF-8 (category Ma: Malformed, bad data)
```

```julia
julia> versioninfo()
Julia Version 1.6.0
Commit f9720dc2eb (2021-03-24 12:55 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, cascadelake)
Environment:
  JULIA_PROJECT = @.
```

(also seems to occur on julia 1.6.1)

**Comment by stevengj** [comment]
Is this an upstream issue in PCRE?  (I'm not seeing this in Julia 1.6.1 on MacOS.)

**Comment by vtjnash** [comment]
Fixed by d336d485466a78826871a7bb1fa2c582730db775 (#39524)

**Comment by vtjnash** [comment]
Should we backport that? I didn't think it was a candidate initially, but perhaps it should be based on encountering it here.
